# HELLO EVERYONE, AGAIN WITH A NEW DATA ANALYSIS ON TELECOM DATASET.

## THIS IS A PART-1 (DATA ANALYSIS) IN THESE WE WILL LEARN EVERYTHING FROM BASICS.


## PART-2 WILL BE UPLOADED TOMORROW WHICH WILL BE BASED ON (DATA VISUALIZATION) WE WILL LEARN ALL IMPORTANT PLOTS USING MATPLOTLIB AND SEABORN.

## Firstly import specific modules:

In [1]:
import numpy as np
import pandas as pd

READ THE CSV FILE:

In [2]:
telecom_data=pd.read_csv('BigML_Dataset_5f4ba58995a9302e500019e2.csv')

In [3]:
telecom_data.head(5)

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [4]:
#LET'S CHECK HOW MANY ROWS AND COLUMNS OUR DATASET CONTAINS:
telecom_data.shape

(3333, 21)

In [5]:
# LET'S CHECK EACH COLUMN NAME:
telecom_data.columns

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')

In [6]:
# LET'S CHECK THE DATATYPE OF EACH COLUMN:
telecom_data.dtypes

state                      object
account length              int64
area code                   int64
phone number               object
international plan         object
voice mail plan            object
number vmail messages       int64
total day minutes         float64
total day calls             int64
total day charge          float64
total eve minutes         float64
total eve calls             int64
total eve charge          float64
total night minutes       float64
total night calls           int64
total night charge        float64
total intl minutes        float64
total intl calls            int64
total intl charge         float64
customer service calls      int64
churn                        bool
dtype: object

Now, we use info() method to gain some information related to our dataset:

In [7]:
telecom_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

AS THESE DATASET HAS NO NULL VALUES,YOU CAN SEE IN ABOVE OUTPUT.

We can change the column type with the astype method. Let’s apply this method to the Churn feature to convert it into int64:

In [8]:
telecom_data['churn']=telecom_data['churn'].astype('int64')

Let's see some statistical information related to our dataset:

In [9]:
telecom_data.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls        churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

In order to see the statistical information of non-numerical datatypes we have to add include parameter.

Now we use value_counts method on categorical feature to see how many unique values our specific column contain:

In [10]:
telecom_data['churn'].value_counts()

0    2850
1     483
Name: churn, dtype: int64

You can see in above output that there is 2850 out of 3333 are loyal.

If you want to see above result in fractional form we had to add normalize parameter = True.

In [11]:
telecom_data['churn'].value_counts(normalize=True).round(2)

0    0.86
1    0.14
Name: churn, dtype: float64

## SORTING OF OUR DATASET.

In [12]:
telecom_data.head(2)

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3                2.7   
1                13.7                 3                3.7   

   customer service calls  churn  
0                       1      0  
1                       1      0  

[2 rows x 21 columns]

In [13]:
telecom_data.sort_values(by=['total day calls'],ascending=False).head()

state  account length  area code phone number international plan  \
1121    MI              91        415     390-7930                 no   
468     AZ              86        415     392-2381                 no   
1460    MT              80        415     361-8288                 no   
2392    WY              90        510     400-8069                 no   
1057    WV              86        415     332-2258                 no   

     voice mail plan  number vmail messages  total day minutes  \
1121              no                      0              154.4   
468              yes                     32               70.9   
1460              no                      0              198.1   
2392              no                      0              125.4   
1057             yes                     38              123.0   

      total day calls  total day charge  ...  total eve calls  \
1121              165             26.25  ...              121   
468               163             12.05  ...              121   
1460              160             33.68  ...               87   
2392              158             21.32  ...               83   
1057              158             20.91  ...              119   

      total eve charge  total night minutes  total night calls  \
1121             14.31                239.9                 81   
468              14.17                244.9                105   
1460             13.32                182.1                 76   
2392             22.87                238.6                103   
1057             11.38                138.2                103   

      total night charge  total intl minutes  total intl calls  \
1121               10.80                11.7                 4   
468                11.02                11.1                 5   
1460                8.19                 9.3                 3   
2392               10.74                11.0                 7   
1057                6.22                13.3                 4   

      total intl charge  customer service calls  churn  
1121               3.16                       5      1  
468                3.00                       3      0  
1460               2.51                       3      0  
2392               2.97                       1      0  
1057               3.59                       1      0  

[5 rows x 21 columns]

### what is the proportion of churned users in our dataframe?

In [14]:
telecom_data['churn'].mean()

0.14491449144914492

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt.

### How much time (on average) do churned users spend on the phone during daytime?

In [15]:
telecom_data.head(2)

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3                2.7   
1                13.7                 3                3.7   

   customer service calls  churn  
0                       1      0  
1                       1      0  

[2 rows x 21 columns]

In [16]:
telecom_data[telecom_data['churn']==1]['total day minutes'].mean()

206.91407867494814

### What is the maximum length of international calls among loyal users (Churn == 0) who do not have an international plan?

In [17]:
telecom_data[(telecom_data['churn']==0)  & (telecom_data['international plan'] =='no')]['total intl minutes'].max()

18.9

## INDEXING:

#### The loc method is used for indexing by name, while iloc() is used for indexing by number.

In [18]:
telecom_data.iloc[0:5,0:3]

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

It shows that give us the values of first five rows for the first three column in above output.

In [19]:
telecom_data[:1] ##IF YOU NEED THE FIRST LINE

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3                2.7   

   customer service calls  churn  
0                       1      0  

[1 rows x 21 columns]

In [20]:
telecom_data[-1:] ##IF YOU WANT LAST LINE

state  account length  area code phone number international plan  \
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
3332              22.6                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
3332               10.86                13.7                 4   

      total intl charge  customer service calls  churn  
3332                3.7                       0      0  

[1 rows x 21 columns]

## APPLYING FUNCTIONS TO ROWS AND COLUMNS:

### TO APPLY THE FUNCTIONS TO EACH COLUMN YOU NEED TO ADD APPLY() FUNCTION: 
    FOR EG: WE NEED TO APPLY MAX FUNCTION ON EACH COLUMN()

In [21]:
telecom_data.apply(np.max)

state                           WY
account length                 243
area code                      510
phone number              422-9964
international plan             yes
voice mail plan                yes
number vmail messages           51
total day minutes            350.8
total day calls                165
total day charge             59.64
total eve minutes            363.7
total eve calls                170
total eve charge             30.91
total night minutes            395
total night calls              175
total night charge           17.77
total intl minutes              20
total intl calls                20
total intl charge              5.4
customer service calls           9
churn                            1
dtype: object

### LAMBDA FUNCTION:They can have any number of arguments but only one expression

FOR EG: IF WE NEED TO SELECT ALL STATES STARTING WITH 'T' USING LAMBDA FUNCTION.

In [22]:
telecom_data[telecom_data['state'].apply(lambda state:state[0]=='T')]

state  account length  area code phone number international plan  \
19      TX              73        415     373-2782                 no   
69      TX             150        510     374-8042                 no   
84      TX             106        510     395-3026                 no   
121     TX              41        408     370-7550                 no   
128     TX              92        408     409-5939                 no   
...    ...             ...        ...          ...                ...   
3168    TX             201        415     415-5476                 no   
3210    TN              93        510     344-6847                yes   
3289    TN              39        408     364-8731                 no   
3295    TN             110        415     391-5516                 no   
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
19                no                      0              224.4   
69                no                      0              178.9   
84                no                      0              210.6   
121              yes                     22              213.8   
128               no                      0               91.7   
...              ...                    ...                ...   
3168              no                      0              225.9   
3210              no                      0              168.4   
3289              no                      0              187.2   
3295              no                      0              227.7   
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
19                 90             38.15  ...               88   
69                101             30.41  ...              110   
84                 96             35.80  ...               85   
121               102             36.35  ...               86   
128                90             15.59  ...              123   
...               ...               ...  ...              ...   
3168              110             38.40  ...               86   
3210              114             28.63  ...              127   
3289              110             31.82  ...              116   
3295               88             38.71  ...               96   
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
19               13.56                192.8                 74   
69               14.37                148.6                100   
84               21.18                191.4                 88   
121              12.05                142.2                123   
128              16.46                175.0                 86   
...                ...                  ...                ...   
3168             25.42                251.3                 81   
3210             23.46                196.2                 48   
3289              9.75                104.7                 83   
3295             14.45                128.7                 57   
3332             22.60                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
19                  8.68                13.0                 2   
69                  6.69                13.8                 3   
84                  8.61                12.4                 1   
121                 6.40                 7.2                 3   
128                 7.88                 9.2                 4   
...                  ...                 ...               ...   
3168               11.31                11.2                 4   
3210                8.83                11.4                 3   
3289                4.71                13.2                 5   
3295                5.79                11.7                 5   
3332               10.86           

IN ABOVE OUTPUT 125 ROWS STARTS WITH STATE = 'T'.

### LET'S CHECK TOTAL DAY CALLS WHICH ARE GREATER THAN 90.

In [23]:
telecom_data[telecom_data['total day calls'] > 90]

state  account length  area code phone number international plan  \
0       KS             128        415     382-4657                 no   
1       OH             107        415     371-7191                 no   
2       NJ             137        415     358-1921                 no   
4       OK              75        415     330-6626                yes   
5       AL             118        510     391-8027                yes   
...    ...             ...        ...          ...                ...   
3326    OH              96        415     347-6812                 no   
3327    SC              79        415     348-3830                 no   
3330    RI              28        510     328-8230                 no   
3331    CT             184        510     364-6381                yes   
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
0                yes                     25              265.1   
1                yes                     26              161.6   
2                 no                      0              243.4   
4                 no                      0              166.7   
5                 no                      0              223.4   
...              ...                    ...                ...   
3326              no                      0              106.6   
3327              no                      0              134.7   
3330              no                      0              180.8   
3331              no                      0              213.8   
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
0                 110             45.07  ...               99   
1                 123             27.47  ...              103   
2                 114             41.38  ...              110   
4                 113             28.34  ...              122   
5                  98             37.98  ...              101   
...               ...               ...  ...              ...   
3326              128             18.12  ...               87   
3327               98             22.90  ...               68   
3330              109             30.74  ...               58   
3331              105             36.35  ...               84   
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
0                16.78                244.7                 91   
1                16.62                254.4                103   
2                10.30                162.6                104   
4                12.61                186.9                121   
5                18.75                203.9                118   
...                ...                  ...                ...   
3326             24.21                178.9                 92   
3327             16.12                221.4                128   
3330             24.55                191.9                 91   
3331             13.57                139.2                137   
3332             22.60                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
0                  11.01                10.0                 3   
1                  11.45                13.7                 3   
2                   7.32                12.2                 5   
4                   8.41                10.1                 3   
5                   9.18                 6.3                 6   
...                  ...                 ...               ...   
3326                8.05                14.9                 7   
3327                9.96                11.8                 5   
3330                8.64                14.1                 6   
3331                6.26                 5.0                10   
3332               10.86           

### MAPPING:THIS METHOD CAN BE USE TO REPLACE A VALUE IN A COLUMN FOR EG:{OLD VALUE:NEW VALUE} AS AN ARGUMENT.

In [24]:
F={'no':False,'yes':True}
telecom_data['international plan']=telecom_data['international plan'].map(F)

In [25]:
telecom_data.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]

Again do these for voice mail plan:

In [26]:
telecom_data['voice mail plan']=telecom_data['voice mail plan'].map(F)

In [27]:
telecom_data.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]

## GROUPING METHOD:

In [28]:
columns_to_use=['total day minutes','total eve minutes','total night minutes']
telecom_data.groupby(['churn'])[columns_to_use].describe(percentiles=[])

total day minutes                                            \
                  count        mean        std  min    50%    max   
churn                                                               
0                2850.0  175.175754  50.181655  0.0  177.2  315.6   
1                 483.0  206.914079  68.997792  0.0  217.6  350.8   

      total eve minutes                                             \
                  count        mean        std   min    50%    max   
churn                                                                
0                2850.0  199.043298  50.292175   0.0  199.6  361.8   
1                 483.0  212.410145  51.728910  70.9  211.3  363.7   

      total night minutes                                              
                    count        mean        std   min     50%    max  
churn                                                                  
0                  2850.0  200.133193  51.105032  23.2  200.25  395.0  
1                   483.0  205.231677  47.132825  47.4  204.80  354.9

### CROSSTAB METHOD: THIS METHOD IS USE TO SEE HOW THE OBSERVATION IN OUR DATASET ARE DISTRIBUTED FOR TWO VARIABLES, FOR EG: LET'S TAKE CHURN AND INTERNATIONAL PLAN.

In [30]:
pd.crosstab(telecom_data['churn'],telecom_data['voice mail plan'],normalize=True).round(2)

voice mail plan  False  True 
churn                        
0                 0.60   0.25
1                 0.12   0.02

You can see in above output that majority of people i.e 60% people are loyal and do not use additional voice mail plan.

## PIVOT TABLE:

IN PIVOT TABLE WE USE THREE PARAMETERS i.e 'VALUES','INDEX','AGGFUNC'.

In [34]:
telecom_data.pivot_table(['total day calls','total eve calls','total night calls'],['area code'],aggfunc='mean')

total day calls  total eve calls  total night calls
area code                                                     
408             100.496420        99.788783          99.039379
415             100.576435       100.503927         100.398187
510             100.097619        99.671429         100.601190

In [36]:
telecom_data.pivot_table(['total day calls','total eve calls','total night calls'],['state'],aggfunc='max')

total day calls  total eve calls  total night calls
state                                                     
AK                 129              147                137
AL                 139              144                152
AR                 145              159                144
AZ                 163              146                138
CA                 150              143                126
CO                 145              137                142
CT                 137              144                158
DC                 147              134                133
DE                 140              138                153
FL                 143              168                146
GA                 137              155                141
HI                 147              157                138
IA                 141              142                144
ID                 151              143                164
IL                 141              150                148
IN                 157              152                151
KS                 144              149                148
KY                 147              154                153
LA                 140              138                131
MA                 158              131                145
MD                 145              156                125
ME                 140              140                157
MI                 165              144                155
MN                 147              142                145
MO                 145              143                150
MS                 156              151                147
MT                 160              146                154
NC                 151              149                135
ND                 144              144                151
NE                 146              147                166
NH                 148              142                139
NJ                 145              147                143
NM                 148              152                134
NV                 150              152                143
NY                 134              170                147
OH                 150              136                152
OK                 151              153                156
OR                 143              141                153
PA                 151              155                142
RI                 134              148                175
SC                 149              155                136
SD                 135              146                143
TN                 145              149                127
TX                 143              143                154
UT                 150              149                151
VA                 142              152                140
VT                 136              142                157
WA                 146              119                148
WI                 141              164                146
WV                 158              138                146
WY                 158              131                156

## ADDING COLUMNS TO A DATAFRAME:

In [46]:
total_calls=telecom_data['total day calls']+telecom_data['total eve calls']+telecom_data['total night calls'] + telecom_data['total intl calls']
telecom_data.insert(loc=len(telecom_data.columns),column='Total calls',value=total_calls)
#we use loc function to define at which position we would like to insert our new column so here telecom_data.column means at the end.

In [47]:
telecom_data.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  ...  total night minutes  total night calls  \
0             45.07  ...                244.7                 91   
1             27.47  ...                254.4                103   
2             41.38  ...                162.6                104   
3             50.90  ...                196.9                 89   
4             28.34  ...                186.9                121   

   total night charge  total intl minutes  total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   total intl charge  customer service calls  churn  Total Calls  Total calls  
0               2.70                       1      0          303          303  
1               3.70                       1      0          332          332  
2               3.29                       0      0          333          333  
3               1.78                       2      0          255          255  
4               2.73                       3      0          359          359  

[5 rows x 23 columns]

In [52]:
telecom_data.sort_values(by=['Total calls'],ascending=False)  ##sorting values of total calls in the descending order.

state  account length  area code phone number  international plan  \
3211    ID             103        415     391-7528               False   
1739    NJ              80        415     330-4978               False   
2263    UT              32        510     370-9563               False   
2796    HI              24        415     329-8788               False   
1370    PA              96        510     359-1441               False   
...    ...             ...        ...          ...                 ...   
2492    SC              73        415     344-9347               False   
349     FL             100        510     420-7823               False   
960     AR               5        415     380-2758               False   
2274    GA              74        510     340-8245               False   
1709    NV              96        408     375-6911               False   

      voice mail plan  number vmail messages  total day minutes  \
3211            False                      0               70.9   
1739            False                      0              268.7   
2263             True                     26              243.5   
2796            False                      0              235.6   
1370            False                      0              179.5   
...               ...                    ...                ...   
2492            False                      0              203.3   
349             False                      0              107.0   
960             False                      0              199.2   
2274             True                     31              249.4   
1709            False                      0              197.7   

      total day calls  total day charge  ...  total eve charge  \
3211              134             12.05  ...             11.43   
1739              120             45.68  ...             25.59   
2263              137             41.40  ...             20.13   
2796              132             40.05  ...              9.85   
1370              125             30.52  ...             13.80   
...               ...               ...  ...               ...   
2492               45             34.56  ...             12.06   
349                63             18.19  ...              8.98   
960               106             33.86  ...             15.92   
2274               70             42.40  ...             17.81   
1709               68             33.61  ...             21.29   

      total night minutes  total night calls  total night charge  \
3211                168.8                164                7.60   
1739                167.0                140                7.52   
2263                173.3                149                7.80   
2796                185.4                136                8.34   
1370                264.5                133               11.90   
...                   ...                ...                 ...   
2492                200.7                 71                9.03   
349                 243.1                 74               10.94   
960                 214.0                 85                9.63   
2274                180.6                 75                8.13   
1709                181.2                 67                8.15   

      total intl minutes  total intl calls  total intl charge  \
3211                12.0                 6               3.24   
1739                 5.8                 1               1.57   
2263                 9.0                 9               2.43   
2796                16.2                 2               4.37   
1370                 6.6                 2               1.78   
...                  ...               ...                ...   
2492                 8.5                 6               2.30   
349                 12.8                 3               3.46   
960                 13.3                 3               3.59   
2274                 9.9                 2         

### DROP:

In [64]:
##IF YOU WANT TO USE DROP JUST SIMPLY TYPE:
telecom_data.drop(['Total calls'],axis=1,inplace=True)

In [65]:
telecom_data.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]

You can see there is no column of total calls.

Now, to delete specific rows:

In [67]:
telecom_data.drop([1,2]) ##you can see 1,2 index values is deleted.

state  account length  area code phone number  international plan  \
0       KS             128        415     382-4657               False   
3       OH              84        408     375-9999                True   
4       OK              75        415     330-6626                True   
5       AL             118        510     391-8027                True   
6       MA             121        510     355-9993               False   
...    ...             ...        ...          ...                 ...   
3328    AZ             192        415     414-4276               False   
3329    WV              68        415     370-3271               False   
3330    RI              28        510     328-8230               False   
3331    CT             184        510     364-6381                True   
3332    TN              74        415     400-4344               False   

      voice mail plan  number vmail messages  total day minutes  \
0                True                     25              265.1   
3               False                      0              299.4   
4               False                      0              166.7   
5               False                      0              223.4   
6                True                     24              218.2   
...               ...                    ...                ...   
3328             True                     36              156.2   
3329            False                      0              231.1   
3330            False                      0              180.8   
3331            False                      0              213.8   
3332             True                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
0                 110             45.07  ...               99   
3                  71             50.90  ...               88   
4                 113             28.34  ...              122   
5                  98             37.98  ...              101   
6                  88             37.09  ...              108   
...               ...               ...  ...              ...   
3328               77             26.55  ...              126   
3329               57             39.29  ...               55   
3330              109             30.74  ...               58   
3331              105             36.35  ...               84   
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
0                16.78                244.7                 91   
3                 5.26                196.9                 89   
4                12.61                186.9                121   
5                18.75                203.9                118   
6                29.62                212.6                118   
...                ...                  ...                ...   
3328             18.32                279.1                 83   
3329             13.04                191.3                123   
3330             24.55                191.9                 91   
3331             13.57                139.2                137   
3332             22.60                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
0                  11.01                10.0                 3   
3                   8.86                 6.6                 7   
4                   8.41                10.1                 3   
5                   9.18                 6.3                 6   
6                   9.57                 7.5                 7   
...                  ...                 ...               ...   
3328               12.56                 9.9                 6   
3329                8.61                 9.6                 4   
3330                8.64                14.1                 6   
3331                6.26                 5.0                10   
3332       

## THIS IS THE ANALYSIS BASED ON PANDAS FROM BASICS I HOPE YOU ALL LIKE IT!!!